### PR TITLE
run test jar

### PR DESCRIPTION
**Pull request by elharo**
@suztomo for discussion

### CODE REVIEW

**Comment by elharo** [comment]
I should try using the maven dependency plugin to dump the dependencies of the test artifact to see what Maven is up to.

**Comment by suztomo** [comment]
With today's troubleshooting an error https://github.com/GoogleCloudPlatform/cloud-opensource-java/pull/1982#issuecomment-816745308, I now think that

- In Yoshi Java libraries, certain artifact provides test utility classes. For example google-cloud-core's test-jar provides BaseSerializationTest. Google-cloud-storage project's test uses the utility class through by declaring test-jar artifact dependency.
- The test-jar artifact _happens to include_ other JUnit test cases, but the JUnit test case classes are not intended to be used in any way.

**Comment by suztomo** [comment]
(You probably read this already) I found the documentation on test-jar.

https://maven.apache.org/guides/mini/guide-attached-tests.html#guide-to-using-attached-tests

> Many times you may want to reuse the tests that you have created for a project in another. For example if you have written foo-core and it contains test code in the ${basedir}/src/test/java it would be useful to package up those compiled tests in a JAR and deploy them for general reuse.

**Comment by elharo** [comment]
consider using `mvn dependency:build-classpath` to find the classpath. Then edit it and run.

**Comment by elharo** [comment]
java -cp /Users/elharo/.m2/repository/com/google/guava/guava/30.1.1-android/guava-30.1.1-android.jar:/Users/elharo/.m2/repository/com/google/guava/failureaccess/1.0.1/failureaccess-1.0.1.jar:/Users/elharo/.m2/repository/com/google/guava/listenablefuture/9999.0-empty-to-avoid-conflict-with-guava/listenablefuture-9999.0-empty-to-avoid-conflict-with-guava.jar:/Users/elharo/.m2/repository/com/google/code/findbugs/jsr305/3.0.2/jsr305-3.0.2.jar:/Users/elharo/.m2/repository/org/checkerframework/checker-compat-qual/2.5.5/checker-compat-qual-2.5.5.jar:/Users/elharo/.m2/repository/com/google/errorprone/error_prone_annotations/2.6.0/error_prone_annotations-2.6.0.jar:/Users/elharo/.m2/repository/com/google/j2objc/j2objc-annotations/1.3/j2objc-annotations-1.3.jar:/Users/elharo/.m2/repository/com/google/http-client/google-http-client/1.39.2/google-http-client-1.39.2.jar:/Users/elharo/.m2/repository/io/opencensus/opencensus-contrib-http-util/0.28.0/opencensus-contrib-http-util-0.28.0.jar:/Users/elharo/.m2/repository/com/google/http-client/google-http-client-jackson2/1.39.2/google-http-client-jackson2-1.39.2.jar:/Users/elharo/.m2/repository/com/fasterxml/jackson/core/jackson-core/2.12.2/jackson-core-2.12.2.jar:/Users/elharo/.m2/repository/com/google/api-client/google-api-client/1.31.4/google-api-client-1.31.4.jar:/Users/elharo/.m2/repository/com/google/oauth-client/google-oauth-client/1.31.5/google-oauth-client-1.31.5.jar:/Users/elharo/.m2/repository/com/google/http-client/google-http-client-gson/1.39.2/google-http-client-gson-1.39.2.jar:/Users/elharo/.m2/repository/com/google/http-client/google-http-client-apache-v2/1.39.2/google-http-client-apache-v2-1.39.2.jar:/Users/elharo/.m2/repository/com/google/apis/google-api-services-storage/v1-rev20210127-1.31.0/google-api-services-storage-v1-rev20210127-1.31.0.jar:/Users/elharo/.m2/repository/com/google/code/gson/gson/2.8.6/gson-2.8.6.jar:/Users/elharo/.m2/repository/com/google/cloud/google-cloud-core/1.94.5/google-cloud-core-1.94.5.jar:/Users/elharo/.m2/repository/com/google/auto/value/auto-value-annotations/1.7.4/auto-value-annotations-1.7.4.jar:/Users/elharo/.m2/repository/com/google/api/grpc/proto-google-common-protos/2.1.0/proto-google-common-protos-2.1.0.jar:/Users/elharo/.m2/repository/com/google/cloud/google-cloud-core-http/1.94.5/google-cloud-core-http-1.94.5.jar:/Users/elharo/.m2/repository/com/google/http-client/google-http-client-appengine/1.39.2/google-http-client-appengine-1.39.2.jar:/Users/elharo/.m2/repository/com/google/api/gax-httpjson/0.80.0/gax-httpjson-0.80.0.jar:/Users/elharo/.m2/repository/com/google/api/gax/1.63.0/gax-1.63.0.jar:/Users/elharo/.m2/repository/com/google/auth/google-auth-library-credentials/0.25.2/google-auth-library-credentials-0.25.2.jar:/Users/elharo/.m2/repository/com/google/auth/google-auth-library-oauth2-http/0.25.2/google-auth-library-oauth2-http-0.25.2.jar:/Users/elharo/.m2/repository/com/google/api/api-common/1.10.1/api-common-1.10.1.jar:/Users/elharo/.m2/repository/javax/annotation/javax.annotation-api/1.3.2/javax.annotation-api-1.3.2.jar:/Users/elharo/.m2/repository/io/opencensus/opencensus-api/0.28.0/opencensus-api-0.28.0.jar:/Users/elharo/.m2/repository/io/grpc/grpc-context/1.37.0/grpc-context-1.37.0.jar:/Users/elharo/.m2/repository/com/google/api/grpc/proto-google-iam-v1/1.0.11/proto-google-iam-v1-1.0.11.jar:/Users/elharo/.m2/repository/com/google/protobuf/protobuf-java/3.15.8/protobuf-java-3.15.8.jar:/Users/elharo/.m2/repository/com/google/protobuf/protobuf-java-util/3.15.8/protobuf-java-util-3.15.8.jar:/Users/elharo/.m2/repository/org/threeten/threetenbp/1.5.0/threetenbp-1.5.0.jar:/Users/elharo/.m2/repository/junit/junit/4.13.2/junit-4.13.2.jar:/Users/elharo/.m2/repository/org/hamcrest/hamcrest-core/1.3/hamcrest-core-1.3.jar:/Users/elharo/.m2/repository/io/grpc/grpc-api/1.37.0/grpc-api-1.37.0.jar:/Users/elharo/.m2/repository/org/codehaus/mojo/animal-sniffer-annotations/1.20/animal-sniffer-annotations-1.20.jar:/Users/elharo/.m2/repository/io/grpc/grpc-netty-shaded/1.37.0/grpc-netty-shaded-1.37.0.jar:/Users/elharo/.m2/repository/io/grpc/grpc-core/1.37.0/grpc-core-1.37.0.jar:/Users/elharo/.m2/repository/com/google/android/annotations/4.1.1.4/annotations-4.1.1.4.jar:/Users/elharo/.m2/repository/io/perfmark/perfmark-api/0.23.0/perfmark-api-0.23.0.jar:/Users/elharo/.m2/repository/io/grpc/grpc-auth/1.37.0/grpc-auth-1.37.0.jar:/Users/elharo/.m2/repository/io/grpc/grpc-stub/1.37.0/grpc-stub-1.37.0.jar:/Users/elharo/.m2/repository/com/google/api/grpc/grpc-google-cloud-kms-v1/0.88.2/grpc-google-cloud-kms-v1-0.88.2.jar:/Users/elharo/.m2/repository/io/grpc/grpc-protobuf/1.37.0/grpc-protobuf-1.37.0.jar:/Users/elharo/.m2/repository/io/grpc/grpc-protobuf-lite/1.37.0/grpc-protobuf-lite-1.37.0.jar:/Users/elharo/.m2/repository/com/google/api/grpc/proto-google-cloud-kms-v1/0.88.2/proto-google-cloud-kms-v1-0.88.2.jar:/Users/elharo/.m2/repository/com/google/api/grpc/grpc-google-iam-v1/1.0.11/grpc-google-iam-v1-1.0.11.jar:/Users/elharo/.m2/repository/com/google/cloud/google-cloud-core/1.94.5/google-cloud-core-1.94.5-tests.jar:/Users/elharo/.m2/repository/com/google/truth/truth/1.1.2/truth-1.1.2.jar:/Users/elharo/.m2/repository/org/checkerframework/checker-qual/3.9.1/checker-qual-3.9.1.jar:/Users/elharo/.m2/repository/org/ow2/asm/asm/9.0/asm-9.0.jar:/Users/elharo/.m2/repository/org/easymock/easymock/3.6/easymock-3.6.jar:/Users/elharo/.m2/repository/org/objenesis/objenesis/2.6/objenesis-2.6.jar:/Users/elharo/.m2/repository/org/mockito/mockito-core/2.28.2/mockito-core-2.28.2.jar:/Users/elharo/.m2/repository/net/bytebuddy/byte-buddy/1.9.10/byte-buddy-1.9.10.jar:/Users/elharo/.m2/repository/net/bytebuddy/byte-buddy-agent/1.9.10/byte-buddy-agent-1.9.10.jar:/Users/elharo/.m2/repository/com/google/cloud/google-cloud-conformance-tests/0.0.11/google-cloud-conformance-tests-0.0.11.jar:/Users/elharo/.m2/repository/com/google/api/grpc/proto-google-cloud-firestore-v1/1.33.0/proto-google-cloud-firestore-v1-1.33.0.jar:/Users/elharo/.m2/repository/com/google/api/grpc/proto-google-cloud-bigtable-v2/1.11.0/proto-google-cloud-bigtable-v2-1.11.0.jar:/Users/elharo/.m2/repository/org/apache/httpcomponents/httpclient/4.5.13/httpclient-4.5.13.jar:/Users/elharo/.m2/repository/commons-logging/commons-logging/1.2/commons-logging-1.2.jar:/Users/elharo/.m2/repository/commons-codec/commons-codec/1.15/commons-codec-1.15.jar:/Users/elharo/.m2/repository/org/apache/httpcomponents/httpmime/4.5.13/httpmime-4.5.13.jar:/Users/elharo/.m2/repository/org/apache/httpcomponents/httpcore/4.4.14/httpcore-4.4.14.jar:google-cloud-storage-1.113.8-tests.jar org.junit.runner.JUnitCore

File: externaltest/pom.xml [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+      <artifactId>junit</artifactId>
+      <version>4.13.2</version>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.mockito</groupId>
+      <artifactId>mockito-core</artifactId>
+      <version>2.28.2</version>
+      <scope>test</scope>
+    </dependency>
```
Comment (suztomo) [inline externaltest/pom.xml:245]: Do we still need this?

Comment (suztomo) [reply externaltest/pom.xml:245]: The new idea is not about the test-jar any more.

**Comment by elharo** [comment]
This approach roughly works. That is the list of dependencies output by dependencies:list is almost sufficient. Needed to add three things:

1. The artifact being tested itself
2. The tests.jar artifacts
3. google-cloud-core test-jar. Not sure why this one didn't get picked up. 

```
  <dependency>
      <groupId>com.google.cloud</groupId>
      <artifactId>google-cloud-storage</artifactId>
      <scope>test</scope>
      <version>1.113.15</version>
      <classifier>tests</classifier>
      <type>test-jar</type>
    </dependency>
  <dependency>
      <groupId>com.google.cloud</groupId>
      <artifactId>google-cloud-storage</artifactId>
      <version>1.113.15</version>
    </dependency>
  <dependency>
      <groupId>com.google.cloud</groupId>
      <artifactId>google-cloud-core</artifactId>
      <scope>test</scope>
      <version>1.94.6</version>
      <classifier>tests</classifier>
      <type>test-jar</type>
    </dependency>
```


One integration test fails due to setup issues:

```
[ERROR] Errors: 
[ERROR] com.google.cloud.storage.it.ITStorageTest.null
[ERROR]   Run 1: ITStorageTest.beforeClass:206 » Storage java.lang.NullPointerException: Requir...
[ERROR]   Run 2: ITStorageTest.afterClass:247 » Execution com.google.cloud.storage.StorageExcep...
```

Tempted to exclude integration tests for now.

**Comment by elharo** [comment]
Tests run: 2, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 0.375 s <<< FAILURE! - in com.google.cloud.storage.it.ITStorageTest
com.google.cloud.storage.it.ITStorageTest  Time elapsed: 0.374 s  <<< ERROR!
com.google.cloud.storage.StorageException: java.lang.NullPointerException: Required parameter project must be specified.
	at com.google.cloud.storage.it.ITStorageTest.beforeClass(ITStorageTest.java:206)
Caused by: java.lang.NullPointerException: Required parameter project must be specified.

**Comment by suztomo** [comment]
That test seems an integration test that would require a real Google Cloud infra.

**Comment by elharo** [comment]
That might have just been local. Looks like all checks passed in CI environment.

**Comment by suztomo** [comment]
Is `externaltest/pom.xml` part of the build?


https://github.com/GoogleCloudPlatform/cloud-opensource-java/pull/2003/checks?check_run_id=2346591143

```
[INFO] Reactor Summary:
[INFO] 
[INFO] Google Cloud Platform Supported Libraries 20.0.1-SNAPSHOT SUCCESS [  2.640 s]
[INFO] Google Cloud Long Term Support BOM 0.1.0-SNAPSHOT .. SUCCESS [  0.016 s]
[INFO] BOM Parent 1.0-SNAPSHOT ............................ SUCCESS [  0.005 s]
[INFO] Cloud Platform Supported Libraries Upper Bounds Check 2.0.0-SNAPSHOT SUCCESS [ 19.397 s]
[INFO] Cloud Tools Open Source Code Hygiene Tooling 1.5.9-SNAPSHOT SUCCESS [ 11.684 s]
[INFO] Maven Dependency Graph Analyzer 1.5.9-SNAPSHOT ..... SUCCESS [10:50 min]
[INFO] integration-tests 1.2.0-SNAPSHOT ................... SUCCESS [04:34 min]
[INFO] Cloud Tools Open Source Code Hygiene Dashboard 1.5.9-SNAPSHOT SUCCESS [01:10 min]
[INFO] Linkage Checker Enforcer Rule 1.5.9-SNAPSHOT ....... SUCCESS [04:43 min]
[INFO] Linkage Monitor 1.5.9-SNAPSHOT ..................... SUCCESS [ 39.973 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
```

**Review by elharo — COMMENTED** [review]
PTAL. After excluding a test that requires GCP project and credentials it all passes.

Next step is to turn this into a script that generates all this from the LTS BOM.

**Review by suztomo — APPROVED** [review]
Approved just in case you need to merge this to master branch to continue the work.

**Comment by suztomo** [comment]
(Year-end cleanup)

**Pull request closed without merging**